doc/other_modules/jenny/runtime/command_storage.md

```diff
@@ -0,0 +1,179 @@
+- The function's arguments must be of types that are known to Yarn: `String`, `num`, `int`,
+  `double`, or `bool`. All arguments must be positional, with no defaults.
+- In order to register the function, use methods `addCommand0()` ... `addCommand3()`, according to
+  the number of function's arguments.
+- If the function's signature has 1 or more booleans at the end, then those arguments will be
+  considered optional and will default to `false`.
+
+
+## Properties
+
+**trueValues**, **falseValues** `Set<String>`
+: The strings that can be recognized as `true`/`false` values respectively.
+
+
+## Methods
+
+**hasCommand**(`String name`) → `bool`
+: Returns the status of whether the command `name` has been added to the storage.
+
+**addCommand0**(`String name`, `FutureOr<void> Function() fn`)
+: Registers a no-argument function `fn` as the command `name`.
+
+**addCommand1**(`String name`, `FutureOr<void> Function(T1) fn`)
+: Registers a single-argument function `fn` as the command `name`.
+
+**addCommand2**(`String name`, `FutureOr<void> Function(T1, T2) fn`)
+: Registers a two-argument function `fn` as the command `name`.
+
+**addCommand3**(`String name`, `FutureOr<void> Function(T1, T2, T3) fn`)
+: Registers a three-argument function `fn` as the command `name`.
+
+**addOrphanedCommand**(`name`)
+: Registers a command `name` which is not backed by any Dart function. Such command will still be
+  delivered to [DialogueView]s via the `onCommand()` callback, but its arguments will not be parsed.
+
+
+## Examples
+
+
+### `<<StartQuest>>`
+
+Suppose we want to have a yarn command `<<StartQuest>>`, which would initiate a quest. The command
+would take the quest name and quest ID as arguments. Technically, just the ID should be enough --
+but then it would be really difficult to read the yarn script and understand what quest is being
+initiated. So, instead we'll pass both the ID and the name, and then check at runtime that the ID
+of the quest matches its name.
+
+A typical invocation of this command might look like this (note that the name of the quest is in
+quotes, otherwise it would be parsed as four different arguments `"Get"`, `"rid"`, `"of"`, and
+`"bandits"`):
+
+```yarn
+<<StartQuest Q037 "Get rid of bandits">>
+```
+
+In order to implement this command, we create a Dart function `startQuest()` with two string
+arguments. The function will do a brief animated "Started quest X" message, but we don't want the
+game to dialogue to wait for that message, so we'll make the function return `void`, not a future.
+Finally, we register the command with `commands.addCommand2()`.
+
+```dart
+class MyGame {
+  late YarnProject yarnProject;
+
+  void startQuest(String questId, String questName) {
+    assert(quests.containsKey(questId));
+    assert(quests[questId]!.name == questName);
+    // ...
+  }
+
+  Future<void> onLoad() async {
+    yarnProject = YarnProject()
+      ..commands.addCommand2('StartQuest', startQuest);
```

Comment:
Is it possible to use generics instead of numbers appended to the functions maybe?

Reply:
AFAIK, Dart doesn't support variadic templates like that (see e.g. https://github.com/dart-lang/language/issues/1774).